core/src/main/java/org/jboss/jandex/EquivalenceKey.java

```diff
@@ -0,0 +1,571 @@
+package org.jboss.jandex;
+
+import java.util.Arrays;
+import java.util.Locale;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * Where the common Jandex classes such as {@link ClassInfo}, {@link MethodInfo} or {@link FieldInfo}
+ * provide strict equality, which includes presence or absence of annotations and comparison of their members,
+ * this class provides <em>equivalence</em>. Two Jandex objects are equivalent if and only if they
+ * denote the same declaration or type, without taking into account any annotations. The prime use case
+ * is to assist with building annotation overlays on top of Jandex.
+ * <p>
+ * An instance of this class, also called an <em>equivalence key</em>, provides 3 fundamental operations,
+ * corresponding to the 3 common methods all Java classes have:
+ * <ul>
+ * <li>{@code equals()}: when two Jandex objects are equivalent, their equivalence keys are equal</li>
+ * <li>{@code hashCode()}: consistent with {@code equals()} described above</li>
+ * <li>{@code toString()}: human readable representation of the equivalence key;
+ * format of the value is not guaranteed and may change without notice</li>
+ * </ul>
+ */
+public class EquivalenceKey {
+    /**
+     * Returns an equivalence key for given class.
+     *
+     * @param clazz the class, may be {@code null}
+     * @return equvalence key for given class, only {@code null} if {@code clazz == null}
+     */
+    public static ClassEquivalenceKey of(ClassInfo clazz) {
```

Comment:
Can we also have `public static EquivalenceKey of(AnnotationTarget target)`?

Reply:
That should be doable, yes. All cases should already be handled anyway.

Reply:
Added.